### PR TITLE
Events emitted from interfaces

### PR DESCRIPTION
**Pull request by bluesign**
draft: events emitted from interfaces

### CODE REVIEW

File: flips/20230417-events-emitted-from-interfaces.md [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+challenges.
+
+## User Benefit
+
+This will establish a more secure environment for the user, ensuring that the
+emitted event is a direct outcome of an action specified within the interface.
+
+## Design Proposal
+
+For instance, a sample extract from the FungibleToken contract can be currently
```
Comment (joshuahannan) [inline flips/20230417-events-emitted-from-interfaces.md:34]:
```suggestion
For instance, a sample extract from the `FungibleToken` contract can be currently
```

File: flips/20230417-events-emitted-from-interfaces.md [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+
+pub contract FungibleToken {
+   
+    pub event TokenWithdrawal(type: Type, amount: UFix64, from: Address?)
+    
+    //Temporary trampoline to emit event 
+    access(contract) fun emitTokenWithdrawal(type: Type, amount: UFix64, from: Address?): Bool{ 
+         emit TokenWithdrawal(type:type, amount:amount, from:from) 
+         return true 
+    }
```
Comment (joshuahannan) [inline flips/20230417-events-emitted-from-interfaces.md:74]: Does it have to be part of a separate function?

**Review by joshuahannan — APPROVED** [review]
I think this is a great idea and I don't see any issues with it! I'm gonna share around to try to get some more feedback

**Comment by bjartek** [comment]
+1

**Review by sisyphusSmiling — APPROVED** [review]
Looks great!

File: flips/20230417-events-emitted-from-interfaces.md (hunk 2 above)
Comment (bluesign) [reply flips/20230417-events-emitted-from-interfaces.md:74]: this is a temporary workaround ( as we don't have a way to emit function from post conditions ) but also it has advantage that we can emit different events depending to parameters.

**Review by turbolent — COMMENTED** [review]
Great idea 👍 

Unless I'm missing something, looking at the "Design Proposal" section, it looks like no actual changes to Cadence are proposed, but rather a pattern. If that is the case, then we do not really need a FLIP, but as mentioned, can describe this pattern in the documentation and recommend it to contract authors.

The "Alternatives Considered" section describes an actual change to Cadence – was the intent to propose that change in the FLIP, or is the idea that such a language change is not necessary, given that the pattern can be used?

If you do want to propose the Cadence change, it might make sense to restructure the proposal and move what is in "Alternatives Considered" to the top of "Design Proposal", and then use the current content of that section (the pattern) as a rational for the proposed feature.

File: flips/20230417-events-emitted-from-interfaces.md [hunk 3]
```diff
@@ -0,0 +1,163 @@
...
+
+## Design Proposal
+
+For instance, a sample extract from the `FungibleToken` contract can be currently
+represented as:
+
+
+```cadence 
+
+pub contract FungibleToken {
```
Comment (turbolent) [inline flips/20230417-events-emitted-from-interfaces.md:40]: Is this declaration missing an `interface` modifier? Currently, the `FungibleToken` standard is an interface.

**Comment by turbolent** [comment]
Just to clarify my understanding:

Is the idea that an event type which is declared in an interface could be refereed to as a concrete type, and thus instantiated, even directly in the interface? e.g.

```cadence
contract interface SomeContractInterface {

    event SomeEvent()

    fun someDefaultFunction() {
        emit SomeEvent()
    }
}
```

Currently the semantics of a type definition nested in an interface is that it is a nested type requirement, i.e. it is not a concrete type, but just declares that the concrete type conforming to the interface must define such a type.

Is the idea to change the behaviour just for events and make them already concrete in the interface?
What about e.g. other kinds of type declarations, like structs and resources.
It might be a bit inconsistent that event type declarations would be concrete, but other kinds of type declarations nested in interfaces would stay nested type requirements, as they are right now.

Maybe asked more generally, does the proposal require/imply the removal of nested type requirements at the same time?

**Comment by joshuahannan** [comment]
I think the proposal would imply the removal of nested type requirements since this is the last thing we need to change to make it okay to remove them. That would make it much cleaner.

I'm not sure for this feature if declaring the events in the resource interface type would make the most sense, or if it would make sense to just force the events to still be declared in the contract instead. I lean more towards declaring them in the contract so it doesn't seem like they are requirements for the interface.

**Comment by bluesign** [comment]
Sorry I was on vacation just saw the comments.

@turbolent I think this should be adopted in core contracts ( FT, NFT etc ) so maybe FLIP can make sense even it is not a cadence change. I started as a draft actually to decide between alternatives, I don't have any idea how hard would it be to allow post conditions to emit events to be honest. 

> Maybe asked more generally, does the proposal require/imply the removal of nested type requirements at the same time?
Yeah you are totally right, I built my prototype on `contract` instead of `contract interface` which requires additional `contract` if we remove nested type requirements. ( which I totally missed ) When thinking about it, other than making events special case, there doesn't seem to be solution. Having a second contract to declare events makes no sense.

**Comment by joshuahannan** [comment]
I'm changing the v2 token standard proposals to be contracts instead of contract interfaces, so you're right that this technically could be done now with separate functions, but I like the idea of being able to emit the events directly from the pre/post conditions, so it would still be nice to make the change to cadence to allow that.

**Comment by turbolent** [comment]
So just to clarify my understanding, is it correct that the FLIP is basically proposing these two things:
- Allow events to be emitted in conditions directly, without the need for a helper function (most important)
- Allow nesting of events in contract interfaces (once/while at the same time removing nested type requirements

**Comment by bluesign** [comment]
> Allow events to be emitted in conditions directly, without the need for a helper function (most important)

I think this one is final decision ( I will update the FLIP ) 

> Allow nesting of events in contract interfaces (once/while at the same time removing nested type requirements)

I think if we will break things and FT / NFT will be contract from now on. I don't think we need to allow nesting of events in contract interfaces.

**Comment by bluesign** [comment]
@SupunS I would appreciate that

**Comment by SupunS** [comment]
Opened a new PR https://github.com/onflow/flips/pull/111 with the updated proposal.

**Comment by turbolent** [comment]
replaced by #111

**Pull request closed without merging**